### ISSUE TITLE
Fetch associations with class name containing hyphen

### ISSUE THREAD

**Issue by dilas** [JsonApiClient/json_api_client#134]
Hello,

I'm having trouble to fetch associations with class name containing hyphen.

My question is: anyone used a model with relations containing hyphen in Ruby?

Reading about the recommendations from JSON API in regard of resource names: http://jsonapi.org/recommendations/ and 
the hyphen minus (U+002D HYPHEN-MINUS, "-") as separator between multiple words is considered safe.

For example (server-side):

``` ruby
class ReportResource < JSONAPI::Resource
  attributes :id, :title, :owner

  has_many :report_sections
end

class ReportSectionResource < JSONAPI::Resource
  attributes  :id, :slug, :name, :active

  has_many :reports
end
```

On client-side:

``` ruby
class Report < JsonApiClient::Resource
end

class ReportSection < JsonApiClient::Resource
end

report = API::Client::Report.where(id: report_id).includes(:'report-sections').first # error occours here
```

```
"/api/reports?filter[id]=1&include=report-sections&page[number]=1&page[size]=1"

/home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/utils.rb:24:in `compute_type': uninitialized constant API::Client::Report::Report-section (NameError)
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/included_data.rb:9:in `block in initialize'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/included_data.rb:8:in `each'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/included_data.rb:8:in `inject'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/included_data.rb:8:in `initialize'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/parsers/parser.rb:97:in `new'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/parsers/parser.rb:97:in `handle_included'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/parsers/parser.rb:18:in `block in parse'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/parsers/parser.rb:8:in `tap'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/parsers/parser.rb:8:in `parse'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/query/requestor.rb:63:in `request'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/query/requestor.rb:26:in `get'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/query/builder.rb:88:in `find'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/query/builder.rb:76:in `to_a'
    from /home/vagrant/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/json_api_client-1.0.1/lib/json_api_client/query/builder.rb:58:in `first'
    from gen-report.rb:59:in `<main>'
```

**Comment by theirishpenguin** [comment]
@dilas  I agree. I am working on a fork to help with this problem. I have it working for my use case  - but I need to tidy it up and perhaps put together a PR.

Here is a branch with the current work in progress on this (it also has some other changes).

https://github.com/chingor13/json_api_client/compare/master...theirishpenguin:penguin/integration

Here are the relevant changes in this area in the above diff...

1) In lib/json_api_client.rb I pull in a json_api_client/configuration file - this allows us to configure "formatters" which abstract over the differences between underscores, hyphens and camelCase in attributes, urls and associations.

2) In the diff, you can see the configuration file that gets pulled in (lib/json_api_client/configuration.rb).

3) I pull in the lib/json_api_client/formatter.rb from the https://github.com/cerebris/jsonapi-resources project. This does the heavy work of abstracting over underscores, hyphens, etc

4) In lib/json_api_client/helpers/dynamic_attributes.rb, lib/json_api_client/included_data.rb, lib/json_api_client/query/builder.rb and lib/json_api_client/resource.rb I then use these methods to abstract away the differences.

As a result this let's me make calls like...

SuperCar.where(id: 21).includes(:super_wheels).all.first.super_wheels #=> gives me back an array of SuperWheels from a server api that exposes and endpoint /api/super-wheels

Does this sound similar to your use case? I would be happy to try an polish up this work and get it into json_api_client if it is of use to more people.

The https://github.com/cerebris/jsonapi-resources project appears to have done some great work in the area of configuration and naming flexibility.

**Comment by smargh** [comment]
This problem needs to be addressed. For teams that use `jsonapi-resources` for their API backend, dashes are enforced. Not being able to consume such an API client-side is frustrating, to say the least. But it looks like @theirishpenguin's MR will do the trick. I hope it gets merged sooner rather than later.

**Comment by JohnSmall** [comment]
I need this as well. I'm integrating with non-Rails APIs which use hyphens rather than underscores.

**Comment by lafeber** [comment]
For anyone interested: my solution for a singular, dashed API was to use this method:

    def self.collection_name
      'process-instance'
    end

**Comment by tavlima** [comment]
#205 fixed this. It should be marked as so and the README should be updated.

TL;DR; something like this should go into your base resource:

```
self.json_key_format = :dasherized_key
self.route_format = :dasherized_route
```

**Comment by senid231** [comment]
@dilas this should give you desired result (as @tavlima suggested)
```ruby
class BaseResource < JSONAPI::Resource
  self.route_format = :dasherized_route
end

class ReportResource < BaseResource
  attributes :id, :title, :owner
  has_many :report_sections
end

class ReportSectionResource < BaseResource
  attributes  :id, :slug, :name, :active
  has_many :reports
end

report = API::Client::Report.where(id: report_id).includes(:report_sections).first
```

**Comment by dilas** [comment]
Yeah, now I can remove the old code for doing the same. Thanks! Great project :)
